### DISCUSSION TITLE
Reading and writing field translations programmatically

### DISCUSSION THREAD

**Discussion by enzedonline** [wagtail/wagtail-localize#362] [Q&A]
How could I find if there is a translated value for a given language and field in page/snippet?
Also, how could I programmatically make a translation for a field? Updating the value directly isn't reflected in the admin, not the best solution.

**Comment by kaedroho** [comment]
The translations are stored in the ``StringTranslation`` model in Wagtail Localize (documented in https://www.wagtail-localize.org/ref/models/translation-memory/). You can edit these programmatically and the changes should be reflected in the admin straight away.

**Reply by enzedonline** [reply]
Thanks Karl,
I'd already had a read through that page, unfortunately, without context or examples, I'm not following how I can reference a field in a page/snippet model and get/set translations from that. 

My goals at the moment are to automate the following

- Create snippet translations and publish (ie as per the Translate button on the snippet menu)
- Synchronize and publish snippet translations (ie as per the Sync Translated Snippets button)
- See if a field in page or snippet has been translated
- Make the translation for that field

One thing I think is really important (which will come once I understand the above) is to translate the slug field if the title field gets translated on certain page types (ie blog and related page types) - expecting content editors to both remember and enter a valid slug field isn't realistic. Similar issue with snippets becoming out of sync because the editor hasn't remembered to propagate changes to the translated snippets.

**Reply by kaedroho** [reply]
Thanks for this!

So I think we could do the following:

 - Implement Python APIs for submitting pages for translation and publishing them, to allow this behaviour to be automated
 - Improve the documentation for programmatically inserting `StringTranslation`. Perhaps add a how-to guide as well
 - Improve the UI for submitting translations after editing the source page (this is already on the backlog, but I'm not exactly sure what the best solution for this one is, other than asking on every save. Any ideas?)

> One thing I think is really important (which will come once I understand the above) is to translate the slug field if the title field gets translated on certain page types (ie blog and related page types) - expecting content editors to both remember and enter a valid slug field isn't realistic.

Editors are required to translate the slug, and they are warned if they forget about it. If you change the title of the source page and update the slug as well, this would normally require the user to retranslate both the title and slug of the translated page. Maybe I've misunderstood your issue here?

**Reply by enzedonline** [reply]
Hi Karl, thanks for your response.

I'm primarily thinking of editors that aren't technically literate. With wagtail trans it's a similar issue that the slug field didn't get translated, and the editor doesn't understand how to create a slug entry. On the default language page, or a single language site, I think most editors like this will never look at the slug field since it's generated automatically from the title. 

Key pages are different of course, but on pages where the slug is a slugified title, I'd like to mirror this behaviour for the translated pages - create a translated slug from the translated title if no translation for the slug already exists. The translated slug could then be amended as is the case with default pages.

Psuedo code version triggered when the title is translated:
```
if not slug.has_translation()
    slug.translate(slugify(title))
```

For the UI when syncing - it would be good to preserve the fields already translated and not affected in the sync. It can cause a lot of unexpected work when all the translated fields are suddenly blown away by a single update.

**Reply by hpoul** [reply]
Is there some example (or a test to look at) in the meantime which programmatically translates snippets/pages?

In particular i'm basically importing data from another source and would like to

1. Create a page and set initial values in the base language
2. Publish that page (which should generate the translated pages?)
3. Override a few strings in the translated pages and publish them as well.

I think 1 and 2 can be simply done by the normal wagtail/django APIs, and this seems to work fine for me. I can also find the translated Pages. But i'm not 100% sure how to work with the `StringTranslation` & co models.

thanks
